core/src/avm2/script.rs

```diff
@@ -0,0 +1,286 @@
+
+#[derive(Clone, Debug, Collect)]
+#[collect(require_static)]
+pub struct CollectWrapper<T>(T);
+
+#[derive(Copy, Clone, Debug, Collect)]
+#[collect(no_drop)]
+pub struct TranslationUnit<'gc>(GcCell<'gc, TranslationUnitData<'gc>>);
+
+/// A loaded ABC file, with any loaded ABC items alongside it.
+///
+/// A `TranslationUnit` is constructed when ABC loading begins, and it stores
+/// all loaded ABC items (classes, methods, and scripts) as they are loaded.
+/// Unit items are loaded lazily and retained in the `TranslationUnit` for
+/// later retrieval.
+///
+/// Loaded versions of ABC items consist of the types `Class`, `Method`, and
+/// `Script`, all of which correspond to their `swf` equivalents, but with
+/// names preloaded. This roughly corresponds to the logical "loading" phase of
+/// ABC execution as documented in the AVM2 Overview. "Linking" takes place by
+/// constructing the appropriate runtime object for that item.
+#[derive(Clone, Debug, Collect)]
+#[collect(no_drop)]
+pub struct TranslationUnitData<'gc> {
+    /// The ABC file that all of the following loaded data comes from.
+    abc: CollectWrapper<Rc<AbcFile>>,
+
+    /// All classes loaded from the ABC's class list.
+    classes: HashMap<u32, GcCell<'gc, Class<'gc>>>,
+
+    /// All methods loaded from the ABC's method list.
+    methods: HashMap<u32, Method<'gc>>,
+
+    /// All scripts loaded from the ABC's scripts list.
+    scripts: HashMap<u32, GcCell<'gc, Script<'gc>>>,
+
+    /// All strings loaded from the ABC's strings list.
+    strings: HashMap<u32, AvmString<'gc>>,
+}
+
+impl<'gc> TranslationUnit<'gc> {
+    pub fn from_abc(abc: Rc<AbcFile>, mc: MutationContext<'gc, '_>) -> Self {
+        Self(GcCell::allocate(
+            mc,
+            TranslationUnitData {
+                abc: CollectWrapper(abc),
+                classes: HashMap::new(),
+                methods: HashMap::new(),
+                scripts: HashMap::new(),
+                strings: HashMap::new(),
+            },
+        ))
+    }
+
+    /// Retrieve the underlying `AbcFile` for this translation unit.
+    pub fn abc(self) -> Rc<AbcFile> {
+        self.0.read().abc.0.clone()
+    }
+
+    /// Load a method from the ABC file and return it's method definition.
+    pub fn load_method(
+        self,
+        method_index: u32,
+        mc: MutationContext<'gc, '_>,
+    ) -> Result<Method<'gc>, Error> {
+        let write = self.0.write(mc);
+        if let Some(method) = write.methods.get(&method_index) {
+            return Ok(method.clone());
+        }
+
+        drop(write);
+
+        let method: Result<Gc<'gc, BytecodeMethod<'gc>>, Error> =
+            BytecodeMethod::from_method_index(self, Index::new(method_index), mc)
+                .ok_or_else(|| "Method index does not exist".into());
+        let method: Method<'gc> = method?.into();
+
+        self.0
+            .write(mc)
+            .methods
+            .insert(method_index, method.clone());
```

Comment:
Feels like we should be able to use the entry API with `or_insert_with`, although I guess the issue is that `from_method_index` does some `txunit.read()`.

Reply:
The timing of when `GcCell` locks are held and released is important here (which is why I used the "wrapped `GcCell` pattern" here). Due to the way traits work, all four of these methods can recurse into one another on the same stack at any time, so I'm being *highly* conservative with how long locks are actually being held.

`BytecodeMethod` *never* needs a write lock on the translation unit, and it only ever reads it to get an ABC. In fact, it reads it three times too many because this used to be `abc_ref` (which caused other double-borrow problems). I'm going to change this particular constructor into taking both a `TranslationUnit` and an `AbcFile`, so it doesn't need to hold either read or write locks and we can use entries.

Reply:
Turns out that won't work, but not for the reason you'd think: `or_insert_with` is infallible. We need to generate a verification error if the method index doesn't exist, which we can't return inside of the `or_insert_with` callable. Ergo, I'll be leaving this as-is.

(This seems to be a problem with a lot of Rust methods that take closures...)